Documentation/ecasignalview.1
.TH "ecatools" "1" "18\&.04\&.2004" "" "Multimedia software" 
.PP 
.SH "NAME" 
ecatools \- audio processing utils based on ecasound
.PP 
.SH "SYNOPSIS" 
\fBecaconvert\fP \&.extension file1 [ file2,\&.\&.\&.,fileN ]
.PP 
\fBecafixdc\fP file1 [ file2,\&.\&.\&.,fileN ]
.PP 
\fBecalength\fP file1 [ file2,\&.\&.\&.,fileN ]
.PP 
\fBecamonitor\fP [host][:port]
.PP 
\fBecanormalize\fP file1 [ file2,\&.\&.\&.,fileN ]
.PP 
\fBecaplay\fP [-dhk] file1 [ file2,\&.\&.\&.,fileN ]
.PP 
\fBecasignalview\fP [ options ] [ input ] [ output ]
.PP 
.SH "DESCRIPTION" 
.PP 
Ecatools are simple command line utils that do various audio
processing tasks\&. Most of them support batch processing\&. They all
use ecasound to do the actual audio processing work\&.
.PP 
\fBECACONVERT\fP
.PP 
A tool for converting a set of files to a common target format\&.
This target format is given as the first command line
argument, and its syntax is \fI\&.ext\fP\&.
.PP 
\fBECAFIXDC\fP
.PP 
A simple command-line tool for fixing DC-offset\&.
.PP 
\fBECALENGTH\fP
.PP 
See the separate ecalength(1) man page\&.
.PP 
\fBECAMONITOR\fP
.PP 
Ecamonitor is a util program that can be used 
to monitor an active ecasound session\&. Ecamonitor 
is able to show information about the current 
chainsetup such as list of inputs, outputs, 
current position, effects, the engine status 
and numerous other things\&.
.PP 
By default ecamonitor will try to connect 
to IP4 address "localhost", to port "2868"\&. 
A different address can be specified when 
starting ecamonitor, for example 
"ecamonitor foo\&.mynetwork\&.com" or
"ecamonitor foo\&.mynetwork\&.com:4500"\&.
.PP 
To enable inbound connections, ecasound has to be 
started with the \fI--daemon\fP option\&. Ecamonitor is 
implemented in Python using the NetECI API\&.
.PP 
\fBECANORMALIZE\fP
.PP 
A simple command-line tool for normalizing sample volume\&. 
Finds out how much input file can be amplified without clipping
and, if necessary, does the amplification\&.
.PP 
\fBECAPLAY\fP
.PP 
Command-line tool that plays input files using the default output 
specified in \fI~/\&.ecasoundrc\fP\&. Note! You can abort ecaplay at any
time by pressing Ctrl-C\&. If you are playing multiple files, this will 
stop the current file and begin playing the next one\&. If you want 
to abort playing immediately instead of skipping to the 
next file, press Ctrl-C twice in short succession (around one
second)\&.
.PP 
.IP "-d:debuglevel" 
Set libecasound debug level\&. See ecasound(1)\&.
.IP 
.IP "-h" 
Print usage information and exit\&.
.IP 
.IP "-k:number" 
Skip \'number\' of files\&.
.IP 
.IP "-o:output_device" 
Use \'output_device\' instead of the default output device 
(as specified in \fI~/\&.ecasoundrc\fP\&.
.IP 
.PP 
\fBECASIGNALVIEW\fP
.PP 
Ecasignalview is meant for monitoring signal amplitude and peak 
statistics\&. It accepts the following options:
.PP 
.IP "input" 
The first option not starting with a \'-\' is interpreted 
as input string\&. This can be a device or a filename\&. All formats
supported by libecasound are accepted\&. Defaults to \'/dev/dsp\'\&.
.IP 
.IP "output" 
Similarly to input, second option not starting with a \'-\'
is interpreted as output\&. Defaults to \'null\'\&.
.IP 
.IP "-b:buffersize" 
Buffersize in sample frames\&. Defaults to 128\&.
.IP 
.IP "-c" 
Enable cumulative mode\&. Counters are not reseted after refresh\&.
\fI*deprecated*\fP
.IP 
.IP "-d" 
Enable debug mode\&.
.IP 
.IP "-f:format_string" 
Specify default audio format\&. See ecasound(1) for details\&.
.IP 
.IP "-r:rate_msec" 
Specify screen refresh rate in milliseconds\&.
.PP 
More complete documentation for ecasignalview can be found 
from "Ecasound User\'s Guide" (see \'\'http://www\&.eca\&.cx/ecasound 
-> Documentation\'\')\&.
.PP 
.SH "ENVIRONMENT" 
.PP 
\fIECASOUND\fP
Path to the ecasound executable\&. Affects most ecatools\&.
.PP 
.SH "FILES" 
.PP 
\fI~/\&.ecasound/ecasoundrc\fP
The default ecasound resource file\&. See ecasoundrc(5)
for details\&.
.PP 
.SH "SEE ALSO" 
.PP 
ecasound (1), ecalength(1), "HTML docs in the Documentation subdirectory"
.PP 
.SH "AUTHOR" 
.PP 
Kai Vehmanen, <kai\&.vehmanen |at| wakkanet\&.fi>
.PP 


NEWS
-----------------------------------------------------------------------
** Ecasound - User-visible changes (NEWS)      (last edited: 311001) **
-----------------------------------------------------------------------

About the version numbers... "vX.Y[.Z][devT]" :

  X = major version  - the overall development status
  Y = minor version  - represents a set of planned features (see TODO);
		       an even number means stable while odd marks 
		       development releases
  Z = revision       - small changes to major.minor version (optional)
		
  devT = dev-release - development releases leading to the next stable
		       version

-----------------------------------------------------------------------

Last officially released version is -*> v2.0.3 *-

????2001 (v2.1dev4)
	- added: eiam - 'cs-option' commmand; although explicit use 
		 of cmd-line options as iactive-mod commands (like 
                 "-i:file.wav") is still supported, "cs-option 
                 -i:file.wav" is now the preferred syntax
	- removed: ecasound FAQ file - it only contains about 	 
		   a pageful of useful text - I moved that
		   precious piece of content to the user's 
		   guide :)
	- changed: uses of '-i' an '-o' replaced with 'ai-add' 
		   and 'ao-add' in ECI documentation and 
		   code examples in 'ecasound/examples'
	- changed: user's and programmer's guides updated
	- changed: bug-alert; major rewrite of ecasound's 
                   internal object factory code
	- changed: eiam - output of aio-register, cop-register,
		   ladspa-register and preset-register commands
	- fixed: multitrack mode was not properly detected if 
		 buffering mode was selected explicitly (-B)
	- fixed: eiam - minor 'cs-edit' bugs fixed
	- fixed: 'make install' failed if srcdir != objdir
	- fixed: didn't work with ALSA +0.9b8 CVS-tree
21102001 (v2.1dev3)
	- note: major changes to effect preset system which 
		break some old preset definitions
	- edi-entries: edi-3
	- added: ecasound-config - --ldflags and --prefix
	- added: engine profiling info (printed to stderr)
	- added: -z:multitrack, -z:nomultitrack for forced enabling
		 and disabling of multitrack-mode
	- added: variable-like preset parameters; it's now possible
		 to bind one preset parameter to multiple chainops'
		 parameters (control multiple settings with one
		 parameter)
	- added: support for describing preset parameters (edi-3)
	- changed: pyecasound - makefile-hack that allows pyecasound
		   to work even though compiled against libecasound 
		   and libkvutils that are not installed to 
		   a directory listed in /etc/ld.so.conf
	- changed: db-subsystem - is put to sleep more 
		   aggressively, which should further decrease 
		   its cpu-load
	- changed: db-subsystem - runs once again under sched_fifo
		   scheduling if -r is enabled
	- changed: ALSA xrun reports are now printed to stderr 
		   instead of ecasound debug-subsystem
	- changed: -B:rt selected as the default bmode for 
		   multitrack-setups (instead of -B:lowlatency)
	- changed: preset definitions in 'effect_presets' changed
		   to take advantage of the preset system improvements
	- fixed: major bugs in interpreting LADSPA plugin 
		 parameter hint flags
	- fixed: a nasty bug in bufmode activation code prevented
		 -kos, -kl, and others from working properly
	- fixed: loop devices didn't work well with -z:db; fixes
		 ecasound testsuite case con_test1-8
	- fixed: memory leak in LADSPA handling code; was triggered
	 	 by recent versions of ecamegapedal
	- fixed: pyecasound - make target 'install-exec-hook' failed
		 if --disable-pyecasound was given to configure
14102001 (v2.1dev2)
	- added: ecaplay - -k, -d and -h options, see 'man ecaplay'
	- added: eiam - ai-get-length-samples, ao-get-length-samples,
		 cs-get-length-samples, cs-set-length-samples [smassy]
	- added: support for variable length buffer's for ALSA 0.9
		 output (required by -ei:xx type effects)
	- changed: tuned prefill-parameters for double-buffering
	- changed: when saving chainsetups, if no parameters 
	           are overridden, buffermode definition (-B) is
		   saved instead of individual parameters
	- changed: yet more changes to the multitrack-mode
	- changed: eiam - output of *-status commands
	- fixed: -r option wasn't saved to chainsetups (used
		 to be a session-level option)
	- fixed: typeselect problems when used for outputs
	- fixed: system lockups with -B:rtlowlatency and 
		 rt-object-only chainsetups
	- fixed: misplaced tab prevented eci.py from working
		 with python 2.1 (nested scopes)
	- fixed: if a monitor track input ended before recording 
                 stopped in multitrack mode and -z:db was enabled, 
		 it continued producing unwanted noise to the
		 monitor output
	- fixed: eiam - ladspa-register output had plugin names 
		 and descriptions mixed
09102001 (v2.1dev1)
	- added: chainsetup parser prints a warning if an unknown
	         buffering mode is given to -B:mode 
	- changed: 'rt' buffering mode is preferred over 
		   'rtlowlatency' if no chain operators are 
		   present, and multitrack mode is not active
        - changed: more improvements to double-buffering
	- fixed: critical bug when running in -B:rtlowlatency 
		 mode (caused system-lockups with ALSA 0.9.x
		 oss-emulation)
	- fixed: explicitly specifying -z:nointbuf had no 
		 effect
	- fixed: setting double-buffer length didn't have 
		 any effect (given either in bmode defaults 
		 or as -z:db,dbsize)
06102001 (v2.1dev0)
	- edi-entries: edi-2, edi-6, edi-10, edi-15
	- added: more recent config.guess and config.sub to 
	         the dist package  [juekawa]
	- added: new feedback-% param to -etd effect, doesn't
		 affect old presets using -etd [smassy]
	- added: ALSA device name 'default' is used if no 
                 other name is given ('-o:alsa' is equivalent
	         to '-o:alsa,default')
	- added: special case for mapping "ecasound file.ecs"
		 to "ecasound -s:file.ecs"
	- added: implementation of edi-6; the -z:db

ecatools/ecasignalview.cpp
// ------------------------------------------------------------------------
// ecasignalview.cpp: A simple command-line tools for monitoring
//                    signal amplitude.
// Copyright (C) 1999-2002 Kai Vehmanen
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307  USA
// ------------------------------------------------------------------------

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <iostream>
#include <string>
#include <vector>

#include <cassert>
#include <cmath>
#include <cstdio>

#include <signal.h>
#include <unistd.h>

#include <kvutils/kvu_com_line.h>
#include <kvutils/kvu_utils.h>
#include <kvutils/kvu_numtostr.h>

#include <eca-control-interface.h>

#include "ecicpp_helpers.h"

#if defined(ECA_USE_NCURSES_H) || defined(ECA_USE_NCURSES_NCURSES_H) || defined(ECA_USE_CURSES_H)
#define ECASV_USE_CURSES 1

#ifdef ECA_USE_NCURSES_H
#include <ncurses.h>
#include <term.h> /* for setupterm() */
#elif ECA_USE_NCURSES_NCURSES_H
#include <ncurses/ncurses.h>
#include <ncurses/term.h> /* for setupterm() */
#else
#include <curses.h>
#include <term.h> /* for setupterm() */
#endif

#endif /* ECA_*CURSES_H */

/**
 * Import namespaces 
 */

using namespace std;

/**
 * Type definitions
 */

struct ecasv_channel_stats {
  double last_peak;
  double drawn_peak;
  double max_peak;
  long int clipped_samples;
};

/**
 * Function declarations
 */

int main(int argc, char *argv[]);
void ecasv_parse_command_line(int argc, char *argv[]);
void ecasv_fill_defaults(void);
std::string ecasv_cop_to_string(ECA_CONTROL_INTERFACE* cop);
void ecasv_output_init(void);
void ecasv_output_cleanup(void);
void ecasv_print_vu_meters(ECA_CONTROL_INTERFACE* eci, std::vector<struct ecasv_channel_stats>* chstats);
void ecasv_update_chstats(std::vector<struct ecasv_channel_stats>* chstats, int ch, double value);
void ecasv_create_bar(double value, int barlen, unsigned char* barbuf);
void ecasv_print_usage(void);
void ecasv_signal_handler(int signum);

/**
 * Static global variables
 */

static const string ecatools_signalview_version = "20021028-5";
static const double ecasv_clipped_threshold_const = 1.0f - 1.0f / 16384.0f;
static const int ecasv_bar_length_const = 32;
static const int ecasv_header_height_const = 10;
static const long int ecasv_rate_default_const = 50;
static const long int ecasv_buffersize_default_const = 128;

static unsigned char ecasv_bar_buffer[ecasv_bar_length_const + 1] = { 0 };
static bool ecasv_enable_debug, ecasv_enable_cumulative_mode;
static long int ecasv_buffersize, ecasv_rate_msec;
static string ecasv_input, ecasv_output, ecasv_format_string;
static int ecasv_chcount = 0;

static ECA_CONTROL_INTERFACE* ecasv_eci_repp = 0;

static sig_atomic_t done = 0;

/**
 * Function definitions
 */

int main(int argc, char *argv[])
{
  struct sigaction es_handler;
  es_handler.sa_handler = ecasv_signal_handler;
  sigemptyset(&es_handler.sa_mask);
  es_handler.sa_flags = 0;

  sigaction(SIGTERM, &es_handler, 0);
  sigaction(SIGINT, &es_handler, 0);
  sigaction(SIGQUIT, &es_handler, 0);
  sigaction(SIGABRT, &es_handler, 0);

  struct sigaction ign_handler;
  ign_handler.sa_handler = SIG_IGN;
  sigemptyset(&ign_handler.sa_mask);
  ign_handler.sa_flags = 0;

  /* ignore the following signals */
  sigaction(SIGPIPE, &ign_handler, 0);
  sigaction(SIGFPE, &ign_handler, 0);

  ecasv_parse_command_line(argc,argv);

  ECA_CONTROL_INTERFACE eci;

  eci.command("cs-add default");
  eci.command("c-add default");
  eci.command("cs-set-param -b:" + kvu_numtostr(ecasv_buffersize));

  if (ecasv_format_string.size() > 0) {
    eci.command("cs-set-audio-format " + ecasv_format_string);
  }
  
  string format;
  if (ecicpp_add_input(&eci, ecasv_input, &format) < 0) return -1;

  cout << "Using audio format -f:" << format << "\n";

  ecasv_chcount = ecicpp_format_channels(format);
  cout << "Setting up " << ecasv_chcount << " separate channels for analysis." << endl;
  
  if (ecicpp_add_output(&eci, ecasv_output, format) < 0) return -1;
  
  ecasv_eci_repp = &eci;

  vector<struct ecasv_channel_stats> chstats;

  eci.command("cop-add -evp");
  eci.command("cop-add -ev");
  if (ecasv_enable_cumulative_mode == true) {
    eci.command("cop-set 2,1,1");
  }

  eci.command("cop-select 1");

  if (ecicpp_connect_chainsetup(&eci, "default") < 0) {
    return -1;
  }

  int secs = 0, msecs = ecasv_rate_msec;
  while(msecs > 999) {
    ++secs;
    msecs -= 1000;
  }

  ecasv_output_init();

  eci.command("start");

  while(! done ) {
    kvu_sleep(secs, msecs * 1000000);
    ecasv_print_vu_meters(&eci, &chstats);
  }

  ecasv_output_cleanup();
#ifdef ECASV_USE_CURSES
  endwin();
#endif
  
  return 0;
}

void ecasv_parse_command_line(int argc, char *argv[])
{
  COMMAND_LINE cline = COMMAND_LINE (argc, argv);
  if (cline.size() == 0 ||
      cline.has("--version") ||
      cline.has("--help") ||
      cline.has("-h")) {
    ecasv_print_usage();
    exit(1);
  }

  ecasv_enable_debug = false;
  ecasv_enable_cumulative_mode = false;
  ecasv_rate_msec = 0; 
  ecasv_buffersize = 0;

  cline.begin();
  cline.next(); // 1st argument
  while (cline.end() != true) {
    string arg = cline.current();
    if (arg.size() > 0) {
      if (arg[0] != '-') {
	if (ecasv_input == "")
	  ecasv_input = arg;
	else
	  if (ecasv_output == "")
	    ecasv_output = arg;
      }
      else {
	string prefix = kvu_get_argument_prefix(arg);
	if (prefix == "b") 
	  ecasv_buffersize = atol(kvu_get_argument_number(1, arg).c_str());
	if (prefix == "c") ecasv_enable_cumulative_mode = true;
	if (prefix == "d") ecasv_enable_debug = true;
	if (prefix == "f")
	  ecasv_format_string = string(arg.begin() + 3, arg.end());
	if (prefix == "r") 
	  ecasv_rate_msec = atol(kvu_get_argument_number(1, arg).c_str());
      }
    }
    cline.next();
  }

  ecasv_fill_defaults();
}

void ecasv_fill_defaults(void)
{
  // ECA_RESOURCES ecarc;

  if (ecasv_input.size() == 0) ecasv_input = "/dev/dsp";
  if (ecasv_output.size() == 0) ecasv_output = "null";
  if (ecasv_buffersize == 0) ecasv_buffersize = ecasv_buffersize_default_const;
  if (ecasv_rate_msec == 0) ecasv_rate_msec = ecasv_rate_default_const;
  if (ecasv_format_string.size() == 0) ecasv_format_string = "s16_le,2,44100,i";

  // ecarc.resource("default-audio-format");
}

string ecasv_cop_to_string(ECA_CONTROL_INTERFACE* eci)
{
  eci->command("cop-status");
  return(eci->last_string());
}

void ecasv_output_init(void)
{
#ifdef ECASV_USE_CURSES
    initscr();
    erase();

    mvprintw(0, 0, "******************************************************\n");
    mvprintw(1, 0, "* ecasignalview v%s (C) 1999-2002 Kai Vehmanen  \n", ecatools_signalview_version.c_str());
    mvprintw(2, 0, "******************************************************\n\n");

    mvprintw(4, 0, "input=\"%s\"\noutput=\"%s\"\naudio_format=\"%s\", refresh_rate_ms='%ld', buffersize='%ld'\n", 
	     ecasv_input.c_str(), ecasv_output.c_str(), ecasv_format_string.c_str(), ecasv_rate_msec, ecasv_buffersize);

    memset(ecasv_bar_buffer, ' ', ecasv_bar_length_const - 4);
    ecasv_bar_buffer[ecasv_bar_length_const - 4] = 0;

    mvprintw(8, 0, "channel-# %s| max-peak  clipped-samples\n", ecasv_bar_buffer);
    mvprintw(9, 0, "----------------------------------------------------------------\n");

    move(12, 0);

    refresh();
#endif
}

void ecasv_output_cleanup(void)
{
#ifdef ECASV_USE_CURSES
    endwin();
#endif

    // FIXME: should be enabled
#if 0    
    if (ecasv_eci_repp != 0) {
      cout << endl << endl << endl;
      ecasv_eci_repp->command("cop-status");
    }
#endif
}

void ecasv_print_vu_meters(ECA_CONTROL_INTERFACE* eci, vector<struct ecasv_channel_stats>* chstats)
{
#ifdef ECASV_USE_CURSES
  for(int n = 0; n < ecasv_chcount; n++) {
    eci->command("copp-select " + kvu_numtostr(n + 1));
    eci->command("copp-get");
    double value = eci->last_float();

    ecasv_update_chstats(chstats, n, value);

    ecasv_create_bar((*chstats)[n].drawn_peak, ecasv_bar_length_const, ecasv_bar_buffer);
    mvprintw(ecasv_header_height_const+n, 0, "Ch-%02d: %s| %.5f       %ld\n", 
	     n + 1, ecasv_bar_buffer, (*chstats)[n].max_peak, (*chstats)[n].clipped_samples);
  }
  move(ecasv_header_height_const + 2 + ecasv_chcount, 0);
  refresh();
#else
  cout << ecasv_cop_to_string(eci) << endl;
#endif
}

void ecasv_update_chstats(vector<struct ecasv_channel_stats>* chstats, int ch, double value)
{
  /* 1. in case a new channel is encoutered */
  if (static_cast<int>(chstats->size()) <= ch) {
    chstats->resize(ch + 1);
  }
  
  /* 2. update last_peak and drawn_peak */
  (*chstats)[ch].last_peak = value;
  if ((*chstats)[ch].last_peak < (*chstats)[ch].drawn_peak) {
    (*chstats)[ch].drawn_peak *= ((*chstats)[ch].last_peak / (*chstats)[ch].drawn_peak);
  }
  else {
    (*chstats)[ch].drawn_peak = (*chstats)[ch].last_peak;
  }

  /* 3. update max_peak */
  if (value > (*chstats)[ch].max_peak) {
    (*chstats)[ch].max_peak = value;
  }

  /* 4. update clipped_samples counter */
  if (value > ecasv_clipped_threshold_const) {
    (*chstats)[ch].clipped_samples++;
  }
}

void ecasv_create_bar(double value, int barlen, unsigned char* barbuf)
{
  int curlen = static_cast<int>(rint(((value / 1.0f) * barlen)));
  for(int n = 0; n < barlen; n++) {
    if (n <= curlen)
      barbuf[n] = '*';
    else
      barbuf[n] = ' ';
  }
}

void ecasv_print_usage(void)
{
  cerr << "****************************************************************************\n";
  cerr << "* ecasignalview, v" << ecatools_signalview_version << "\n";
  cerr << "* (C) 1999-2002 Kai Vehmanen, released under GPL licence\n";
  cerr << "****************************************************************************\n";

  cerr << "\nUSAGE: ecasignalview [options] [input] [output] \n";
  cerr << "\toptions:\n";
  cerr << "\t\t-b:buffersize\n";
  // cerr << "\t\t-c (cumulative mode)\n";
  cerr << "\t\t-d (debug mode)\n";
  cerr << "\t\t-f:bits,channels,samplerate\n";
  cerr << "\t\t-r:refresh_msec\n\n";
}

void ecasv_signal_handler(int signum)
{
    cerr << "Interrupted... cleaning up.\n";
    done=1;
}
